libvast/src/segment_store.cpp

```diff
@@ -177,27 +170,135 @@ std::unique_ptr<store::lookup> segment_store::extract(const ids& xs) const {
 }
 
 caf::error segment_store::erase(const ids& xs) {
-  // TODO: implement me
+  VAST_TRACE(VAST_ARG(xs));
+  // Get affected segments.
+  std::vector<uuid> candidates;
+  if (auto err = select_segments(xs, candidates))
+    return err;
+  if (candidates.empty())
+    return caf::none;
+  // Predicate for checking whether `xs` contains all IDs of `ys`.
+  auto is_subset_of_xs = [&](const ids& ys) {
+    // Compare using rank(), because the bitmaps might have different sizes.
+    return contains(xs, ys);
+  };
+  // Counts number of total erased events for user-facing output.
+  uint64_t erased_events = 0;
+  // Convenience function for updating given segment by pruning all IDs in `xs`
+  // from it.
+  // @returns `true` if the entire segment got erased, `false` otherwise.
+  auto update = [&](auto& sref) {
+    auto slice_ids = sref.meta().slice_ids();
+    // Check whether we can drop the entire segment.
+    if (std::all_of(slice_ids.begin(), slice_ids.end(), is_subset_of_xs)) {
+      erased_events += drop(sref);
+      return;
+    }
+    auto segment_id = sref.id();
+    // Get all slices in the segment and generate a new segment that contains
+    // only what's left after dropping the selection.
+    auto segment_ids = sref.meta().flat_slice_ids();
+    std::vector<table_slice_ptr> slices;
+    if (auto maybe_slices = sref.lookup(segment_ids)) {
+      slices = std::move(*maybe_slices);
+      if (slices.empty()) {
+        VAST_WARNING(this, "got no slices after lookup for segment", segment_id,
+                     "=> erases entire segment!");
+        erased_events += drop(sref);
+        return;
+      }
+    } else {
+      VAST_WARNING(this, "was unable to get table slice for segment",
+                   segment_id, "=> erases entire segment!");
+      erased_events += drop(sref);
+      return;
+    }
+    VAST_ASSERT(slices.size() > 0);
+    auto keep_mask = ~xs;
```

Comment:
Can you add a short (single-line suffices) description of the algorithm you're implementing below using `keep_mask`?